### ISSUE TITLE
raspberry pi 2 build error

### ISSUE THREAD

**Issue by eokeeffe** [tangrams/tangram-es#235]
Heya, trying to compile the package and I'm getting an error originating from 
core/textbuffer.cpp
it goes into vcos_types.h saying expected ")" before sizeof statement on lines 151 and 201

had a look at the statements in that header file and they don't have any errors
gcc is version 4.9 and g++ is version 4.9, the CC and CXX were correctly set.

I followed all the build instructions to the letter, just wondering if this could be caused by a variable 
not being set or something

**Comment by blair1618** [comment]
Hey thanks for the heads up, can you post the full error message you got from the compiler?

**Comment by eokeeffe** [comment]
Building CXX object core/CMakeFiles/core.dir/src/text/textBuffer.cpp.o
In file included from /opt/vc/include/interface/vcos/vcos_assert.h:149:0,
                 from /opt/vc/include/interface/vcos/vcos.h:114,
                 from /opt/vc/include/interface/vmcs_host/vc_dispmanx.h:33,
                 from /opt/vc/include/bcm_host.h:50,
                 from /home/pi/tangram-es/core/src/gl.h:31,
                 from /home/pi/tangram-es/core/src/text/textBuffer.h:3,
                 from /home/pi/tangram-es/core/src/text/textBuffer.cpp:1:
/opt/vc/include/interface/vcos/vcos_types.h:201:21: error: expected
unqualified-id before 'sizeof'
 #define countof(x) (sizeof((x)) / sizeof((x)[0]))
                     ^
/home/pi/tangram-es/core/include/glm/glm/detail/setup.hpp:1009:25: note: in
expansion of macro 'countof'
   constexpr std::size_t countof(T const (&)[N])
                         ^
/opt/vc/include/interface/vcos/vcos_types.h:201:21: error: expected ')'
before 'sizeof'
 #define countof(x) (sizeof((x)) / sizeof((x)[0]))
                     ^
/home/pi/tangram-es/core/include/glm/glm/detail/setup.hpp:1009:25: note: in
expansion of macro 'countof'
   constexpr std::size_t countof(T const (&)[N])
                         ^
core/CMakeFiles/core.dir/build.make:261: recipe for target
'core/CMakeFiles/core.dir/src/text/textBuffer.cpp.o' failed
make[3]: *** [core/CMakeFiles/core.dir/src/text/textBuffer.cpp.o] Error 1
make[3]: Leaving directory '/home/pi/tangram-es/build/rpi'
CMakeFiles/Makefile2:120: recipe for target 'core/CMakeFiles/core.dir/all'
failed
make[2]: *** [core/CMakeFiles/core.dir/all] Error 2
make[2]: Leaving directory '/home/pi/tangram-es/build/rpi'
Makefile:113: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/pi/tangram-es/build/rpi'
Makefile:169: recipe for target 'rpi' failed
make: *** [rpi] Error 2


On 2 August 2015 at 19:56, Matt Blair <notifications@github.com> wrote:

> Hey thanks for the heads up, can you post the full error message you got
> from the compiler?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tangrams/tangram-es/issues/235#issuecomment-127054198>
> .
>



-- 
*Forti Et Fideli , Nihil difficile*

**Comment by eokeeffe** [comment]
Good news, I found a solution here and also did a fix for installing the package for Ubuntu Mate rpi2.

First. I went into /opt/vc/include/interfaces/vcos_types.h and commented out the #define countof
there was a comment beneath it saying it was for backwards compatibility anyway.
That solved the problem for the raspbian image.

Second. For Ubuntu Mate after adding the /opt/vc/include folder from raspberrpy/firmware, I changed the cmake to look at /usr/lib instead of /opt/vc/lib , changed -lpthread to -pthread and added the libraries for linking down in the target_link_libraries. Works a treat and the program is running fine.

I'll clean up the fixes and add a declaration for UbuntuMate and raspbian in the raspberrpy.cmake file as well as the patched vcos_type.h file for this project.
If you want a separate branch for UbuntuMate it's no problem
